### PR TITLE
[SPARK-32902][SQL] Logging plan changes for AQE

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

Recently, we added code to log plan changes in the preparation phase in `QueryExecution` for execution (https://github.com/apache/spark/pull/29544). This PR intends to apply the same fix  for logging plan changes in AQE.

### Why are the changes needed?

Easy debugging for AQE plans

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added unit tests.

### CODE REVIEW

**Comment by maropu** [comment]
cc: @cloud-fan @HyukjinKwon

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 1]
```diff
@@ -231,7 +232,9 @@ case class AdaptiveSparkPlanExec(
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:232]: qq: should we remove `spark.sql.adaptive.logLevel`, @cloud-fan, @maryannxue, and @maropu?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:232]: I think they are different. `spark.sql.adaptive.logLevel` controls the logging of plan changes for each AQE re-optimization round. While this PR is to log plan change before/after each rule.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 2]
```diff
@@ -231,7 +232,9 @@ case class AdaptiveSparkPlanExec(
 
       // Run the final plan when there's no more unfinished stages.
       currentPhysicalPlan = applyPhysicalRules(
-        result.newPlan, queryStageOptimizerRules ++ postStageCreationRules)
+        "AQE Physical Plan Optimization",
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:235]: `AQE final query stage optimization`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 3]
```diff
@@ -413,19 +416,24 @@ case class AdaptiveSparkPlanExec(
   }
 
   private def newQueryStage(e: Exchange): QueryStageExec = {
-    val optimizedPlan = applyPhysicalRules(e.child, queryStageOptimizerRules)
+    val optimizedPlan = applyPhysicalRules(
+      "AQE Physical Plan Optimization", e.child, queryStageOptimizerRules)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:420]: `AQE query stage optimization`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 4]
```diff
@@ -534,7 +542,8 @@ case class AdaptiveSparkPlanExec(
     logicalPlan.invalidateStatsCache()
     val optimized = optimizer.execute(logicalPlan)
     val sparkPlan = context.session.sessionState.planner.plan(ReturnAnswer(optimized)).next()
-    val newPlan = applyPhysicalRules(sparkPlan, preprocessingRules ++ queryStagePreparationRules)
+    val newPlan = applyPhysicalRules(
+      "AQE Preparations", sparkPlan, preprocessingRules ++ queryStagePreparationRules)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:546]: `AQE replanning`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 5]
```diff
@@ -630,8 +639,18 @@ object AdaptiveSparkPlanExec {
   /**
    * Apply a list of physical operator rules on a [[SparkPlan]].
    */
-  def applyPhysicalRules(plan: SparkPlan, rules: Seq[Rule[SparkPlan]]): SparkPlan = {
-    rules.foldLeft(plan) { case (sp, rule) => rule.apply(sp) }
+  def applyPhysicalRules(
+      batchName: String,
+      plan: SparkPlan,
+      rules: Seq[Rule[SparkPlan]]): SparkPlan = {
+    val planChangeLogger = new PlanChangeLogger[SparkPlan]()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:646]: shall we create `PlanChangeLogger` instance only once? (keep it as a class member)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala (hunk 2 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala]: NOTE: I uppercased the first letters.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala (hunk 1 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:234]: Looking around the code, logging granularity looks different as @cloud-fan said. A bit confusing though, I think its okay as it is.

**Comment by cloud-fan** [comment]
retest this please

**Review by cloud-fan — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by maropu** [comment]
Thanks, @cloud-fan @viirya ! Merged to master.

**Pull request closed without merging**